README.md
Private Docker Registry deployed with BOSH
==========================================

Run your own private Docker Registry in standalone mode (without requiring the public index) on AWS, OpenStack or vSphere.

For example, tagging and pushing the public `ubuntu:13.04` repository to an AWS-deployed Docker repository:

```
$ docker tag ubuntu:13.04 ec2-54-80-246-141.compute-1.amazonaws.com:5000/ubuntu-1304
$ docker push ec2-54-80-246-141.compute-1.amazonaws.com:5000/ubuntu-1304
The push refers to a repository [ec2-54-80-246-141.compute-1.amazonaws.com:5000/ubuntu-1304] (len: 1)
Sending image list
Pushing repository ec2-54-80-246-141.compute-1.amazonaws.com:5000/ubuntu-1304 (1 tags)
511136ea3c5a: Image successfully pushed
f323cf34fd77: Image successfully pushed
eb601b8965b8: Pushing [==================>                                ] 68.77 MB/181.2 MB 4m1s
```

Usage
-----

To use this BOSH release, first upload it to your bosh:

```
bosh upload release https://bosh.io/d/github.com/cloudfoundry-community/docker-registry-boshrelease
```

To deploy it you will need the source repository that contains templates:

```
git clone https://github.com/cloudfoundry-community/docker-registry-boshrelease.git
cd docker-registry-boshrelease
```

To use BOSH v2 manifests, have a look at `templates/docker-registry.yml`, otherwise
create the manifests using the traditional way with `templates/make_manifest`

For AWS EC2, create a single VM:

```
templates/make_manifest aws-ec2
bosh -n deploy
```

For developers of the release using [bosh-lite](https://github.com/cloudfoundry/bosh-lite), you can quickly create a deployment manifest and deploy:

```
templates/make_manifest warden
bosh -n deploy
```

#### Override security groups

For AWS & Openstack, the default deployment assumes there is a `default` security group. If you wish to use a different security group(s) then you can pass in additional configuration when running `make_manifest` above.

Create a file `my-networking.yml`:

```yaml
---
networks:
  - name: docker-registry1
    type: dynamic
    cloud_properties:
      security_groups:
        - docker_registry
```

Where `- docker_registry` means you wish to use an existing security group called `docker_registry`.

You now suffix this file path to the `make_manifest` command:

```
templates/make_manifest openstack-nova my-networking.yml
bosh -n deploy
```

Access restrictions
-----------------------

The registry can be available for administrative and reading purposes on different rules:
- for **administrative and reading** purposes:
  - default port: 443 (`docker.proxy.port`)
  - default authentication: basic
- you can enable **readonly access** without authentication by setting `docker.proxy.only_auth_for_admin=false`:
  - default port: 444 (`docker.proxy.readonly_port`)
  - default authentication: none

#### Basic authentication

You have to create the new password just like for `.htpasswd` file:
```
htpasswd -nb docker my-very-unique-password
```

Update the `manifests/docker-registry.yml` with the new password hash:

```
instance_groups:
  - name: docker-registry
    properties:
      docker:
        proxy:
          auth_basic:
            docker: "$apr1$EyknLHps$aOq105nTuANVeSOfl/Pla1"
```


SSL settings for Docker
-----------------------

### Signed Certificate from a CA

If you have a public domain (even if it's for internal use) you can get a free certificate from "Let’s Encrypt":
- automated way: https://letsencrypt.org/
- manual way: https://www.sslforfree.com/

Update the `manifests/docker-registry.yml` with the SSL certificate:

```
instance_groups:
  - name: docker-registry
    properties:
      docker:
        proxy:
          ssl:
            cert: |+
              -----BEGIN CERTIFICATE-----
              ... certificate.crt ...
              -----END CERTIFICATE-----
              -----BEGIN CERTIFICATE-----
              ... ca_bundle.crt ...
              -----END CERTIFICATE-----
            key: |+
              -----BEGIN PRIVATE KEY-----
              ... private.key ...
              -----END PRIVATE KEY-----
```


### Self-signed Certificate 

Getting docker working with a private registry can be a time-consuming
task when you use a self-signed certificate. Those are the steps to make it easy (but also insecure):

```
# Define the IP of the Docker registry server. You could use a DNS name
# but then you have to modify the settings of the file below
REGISTRY_IP="assign the ip of the registry"

# Create a folder for the process
mkdir ssl

# Create the configuration file for openssl settings
cat <<EOF > ssl/insecure.cnf
[req]
distinguished_name = req_distinguished_name
x509_extensions = v3_req
prompt = no

[req_distinguished_name]
C = NL
ST = Holland
L = Dordrecht
O = Springer Nature
CN = *

[v3_req]
subjectKeyIdentifier = hash
authorityKeyIdentifier = keyid,issuer
basicConstraints = CA:TRUE
subjectAltName = @alt_names

[alt_names]
IP = $REGISTRY_IP
DNS.1 = *
DNS.2 = *.*
DNS.3 = *.*.*
DNS.4 = *.*.*.*
DNS.5 = *.*.*.*.*
DNS.6 = *.*.*.*.*.*
DNS.7 = *.*.*.*.*.*.*

EOF

# Run openssl with the previous settings to create the self-signed certificate
openssl req -x509 -batch -nodes -newkey rsa:2048 -keyout ssl/registry.key -out ssl/registry.cert -config ssl/insecure.cnf -days 3650
```

Assign the content of the file `ssl/registry.key` to
`docker.proxy.ssl.key` property and the content of `ssl/registry.cert` to `docker.proxy.ssl.cert`. Re-deploy if it is needed.

Now it is time to configure the local docker daemon to accept the
new Registry certificates. Docker daemon uses the system certificates,
so we have to include the CA certificate there (On Ubuntu):

```
# Copy CA cert to system certificates folder
sudo cp ssl/registry.cert /usr/share/ca-certificates/
# Enable the new CA
sudo update-ca-certificates
# Restart Docker
sudo service docker restart
```

Now you can push an image to the private Docker Registry:

```
# docker push 10.230.30.76:443/bosh
The push refers to a repository [10.230.30.76:443/bosh]
5f70bf18a086: Pushed
587d5d1f5fec: Pushed
dc5fe436f9a6: Pushed
82cfb997c4aa: Pushed
a96da73480aa: Pushed
8d978c6d6037: Pushed
3fb897e02c04: Pushed
bb1d2c69066a: Pushed
5ee9df15e635: Pushed
14d1e894d916: Pushed
fe17e7c4c5c1: Pushed
3eb2e43b8463: Pushed
69e57e993b01: Pushed
e678c68d0466: Pushed
6333a97cf74a: Pushed
44599e8fdbd0: Pushed
c8bff16118ee: Pushed
62f86f3d6a38: Pushed
b4f99d457706: Pushed
f1869edf32d7: Pushed
10ee77e0916d: Pushed
0050a91818d4: Pushed
cba62530f74a: Pushed
d7a905e2ee84: Pushed
a739ea00a948: Pushed
470641744213: Pushed
latest: digest: sha256:74ca3932f4cda9b1bf31b38518e35b8cc52cda3fa6ec767dacacafc2fb27f690 size: 7208
```

Links with more information
 * https://github.com/docker/docker/issues/8849
 * https://github.com/docker/distribution/issues/1731


ci/release_notes.md
# Bug Fixes

- Create the local `nfs_server.share_path` if it doesn't already
  exist, which may be the case if it's being overridden.

- Replace *ALL* occurrences of `$VCAP_UID` and `$VCAP_GID`
  variable interpolation sites in the generated /etc/exports.


src/common/utils.sh

mkdir -p /var/vcap/sys/log

exec >>/var/vcap/sys/log/nfs.log 2>&1

pid_guard() {
  echo "------------ STARTING `basename $0` at `date` --------------" | tee /dev/stderr
  pidfile=$1
  name=$2

  if [ -f "$pidfile" ]; then
    pid=$(head -1 "$pidfile")

    if [ -n "$pid" ] && [ -e /proc/$pid ]; then
      echo "$name is already running, please stop it first"
      exit 1
    fi

    echo "Removing stale pidfile..."
    rm $pidfile
  fi
}

wait_pidfile() {
  pidfile=$1
  try_kill=$2
  timeout=${3:-0}
  force=${4:-0}
  countdown=$(( $timeout * 10 ))

  if [ -f "$pidfile" ]; then
    pid=$(head -1 "$pidfile")

    if [ -z "$pid" ]; then
      echo "Unable to get pid from $pidfile"
      exit 1
    fi

    if [ -e /proc/$pid ]; then
      if [ "$try_kill" = "1" ]; then
        echo "Killing $pidfile: $pid "
        kill $pid
      fi
      while [ -e /proc/$pid ]; do
        sleep 0.1
        [ "$countdown" != '0' -a $(( $countdown % 10 )) = '0' ] && echo -n .
        if [ $timeout -gt 0 ]; then
          if [ $countdown -eq 0 ]; then
            if [ "$force" = "1" ]; then
              echo -ne "\nKill timed out, using kill -9 on $pid... "
              kill -9 $pid
              sleep 0.5
            fi
            break
          else
            countdown=$(( $countdown - 1 ))
          fi
        fi
      done
      if [ -e /proc/$pid ]; then
        echo "Timed Out"
      else
        echo "Stopped"
      fi
    else
      echo "Process $pid is not running"
    fi

    rm -f $pidfile
  else
    echo "Pidfile $pidfile doesn't exist"
  fi
}

kill_and_wait() {
  pidfile=$1
  # Monit default timeout for start/stop is 30s
  # Append 'with timeout {n} seconds' to monit start/stop program configs
  timeout=${2:-25}
  force=${3:-1}

  wait_pidfile $pidfile 1 $timeout $force
}

running_in_container() {
  grep -q '/instance' /proc/self/cgroup
}


jobs/docker_registry/templates/helpers/ctl_setup.sh
#!/usr/bin/env bash

# Setup env vars and folders for the ctl script
# This helps keep the ctl script as readable
# as possible

# Usage options:
# source /var/vcap/jobs/foobar/helpers/ctl_setup.sh JOB_NAME OUTPUT_LABEL
# source /var/vcap/jobs/foobar/helpers/ctl_setup.sh foobar
# source /var/vcap/jobs/foobar/helpers/ctl_setup.sh foobar foobar
# source /var/vcap/jobs/foobar/helpers/ctl_setup.sh foobar nginx

set -e # exit immediately if a simple command exits with a non-zero status
set -u # report the usage of uninitialized variables

JOB_NAME=$1
output_label=${1:-JOB_NAME}

export JOB_DIR=/var/vcap/jobs/$JOB_NAME
chmod 755 $JOB_DIR # to access file via symlink

# Load some bosh deployment properties into env vars
# Try to put all ERb into data/properties.sh.erb
# incl $NAME, $JOB_INDEX, $WEBAPP_DIR
source $JOB_DIR/data/properties.sh

source $JOB_DIR/helpers/ctl_utils.sh
redirect_output ${output_label}

export HOME=${HOME:-/home/vcap}

# Add all packages' /bin & /sbin into $PATH
for package_bin_dir in $(ls -d /var/vcap/packages/*/*bin)
do
  export PATH=${package_bin_dir}:$PATH
done

export LD_LIBRARY_PATH=${LD_LIBRARY_PATH:-''} # default to empty
for package_bin_dir in $(ls -d /var/vcap/packages/*/lib)
do
  export LD_LIBRARY_PATH=${package_bin_dir}:$LD_LIBRARY_PATH
done

# Setup log, run and tmp folders

export RUN_DIR=/var/vcap/sys/run/$JOB_NAME
export LOG_DIR=/var/vcap/sys/log/$JOB_NAME
export TMP_DIR=/var/vcap/sys/tmp/$JOB_NAME
export STORE_DIR=/var/vcap/store/$JOB_NAME
for dir in $RUN_DIR $LOG_DIR $TMP_DIR $STORE_DIR
do
  mkdir -p ${dir}
  chown vcap:vcap ${dir}
  chmod 775 ${dir}
done
export TMPDIR=$TMP_DIR

export C_INCLUDE_PATH=/var/vcap/packages/mysqlclient/include/mysql:/var/vcap/packages/sqlite/include:/var/vcap/packages/libpq/include
export LIBRARY_PATH=/var/vcap/packages/mysqlclient/lib/mysql:/var/vcap/packages/sqlite/lib:/var/vcap/packages/libpq/lib

for profile in $(ls /var/vcap/packages/*/profile.sh)
do
	source $profile
done

PIDFILE=$RUN_DIR/$JOB_NAME.pid

echo '$PATH' $PATH


jobs/cache/templates/helpers/ctl_utils.sh
# Helper functions used by ctl scripts

# links a job file (probably a config file) into a package
# Example usage:
# link_job_file_to_package config/redis.yml [config/redis.yml]
# link_job_file_to_package config/wp-config.php wp-config.php
link_job_file_to_package() {
  source_job_file=$1
  target_package_file=${2:-$source_job_file}
  full_package_file=$WEBAPP_DIR/${target_package_file}

  link_job_file ${source_job_file} ${full_package_file}
}

# links a job file (probably a config file) somewhere
# Example usage:
# link_job_file config/bashrc /home/vcap/.bashrc
link_job_file() {
  source_job_file=$1
  target_file=$2
  full_job_file=$JOB_DIR/${source_job_file}

  echo link_job_file ${full_job_file} ${target_file}
  if [[ ! -f ${full_job_file} ]]
  then
    echo "file to link ${full_job_file} does not exist"
  else
    # Create/recreate the symlink to current job file
    # If another process is using the file, it won't be
    # deleted, so don't attempt to create the symlink
    mkdir -p $(dirname ${target_file})
    ln -nfs ${full_job_file} ${target_file}
  fi
}

# If loaded within monit ctl scripts then pipe output
# If loaded from 'source ../utils.sh' then normal STDOUT
redirect_output() {
  SCRIPT=$1
  mkdir -p /var/vcap/sys/log/monit
  exec 1>> /var/vcap/sys/log/monit/$SCRIPT.log 2>&1
}

pid_guard() {
  pidfile=$1
  name=$2

  if [ -f "$pidfile" ]; then
    pid=$(head -1 "$pidfile")

    if [ -n "$pid" ] && [ -e /proc/$pid ]; then
      echo "$name is already running, please stop it first"
      exit 1
    fi

    echo "Removing stale pidfile..."
    rm $pidfile
  fi
}

wait_pid() {
  pid=$1
  try_kill=$2
  timeout=${3:-0}
  force=${4:-0}
  countdown=$(( $timeout * 10 ))

  echo wait_pid $pid $try_kill $timeout $force $countdown
  if [ -e /proc/$pid ]; then
    if [ "$try_kill" = "1" ]; then
      echo "Killing $pidfile: $pid "
      kill $pid
    fi
    while [ -e /proc/$pid ]; do
      sleep 0.1
      [ "$countdown" != '0' -a $(( $countdown % 10 )) = '0' ] && echo -n .
      if [ $timeout -gt 0 ]; then
        if [ $countdown -eq 0 ]; then
          if [ "$force" = "1" ]; then
            echo -ne "\nKill timed out, using kill -9 on $pid... "
            kill -9 $pid
            sleep 0.5
          fi
          break
        else
          countdown=$(( $countdown - 1 ))
        fi
      fi
    done
    if [ -e /proc/$pid ]; then
      echo "Timed Out"
    else
      echo "Stopped"
    fi
  else
    echo "Process $pid is not running"
    echo "Attempting to kill pid anyway..."
    kill $pid
  fi
}

wait_pidfile() {
  pidfile=$1
  try_kill=$2
  timeout=${3:-0}
  force=${4:-0}
  countdown=$(( $timeout * 10 ))

  if [ -f "$pidfile" ]; then
    pid=$(head -1 "$pidfile")
    if [ -z "$pid" ]; then
      echo "Unable to get pid from $pidfile"
      exit 1
    fi

    wait_pid $pid $try_kill $timeout $force

    rm -f $pidfile
  else
    echo "Pidfile $pidfile doesn't exist"
  fi
}

kill_and_wait() {
  pidfile=$1
  # Monit default timeout for start/stop is 30s
  # Append 'with timeout {n} seconds' to monit start/stop program configs
  timeout=${2:-25}
  force=${3:-1}
  if [[ -f ${pidfile} ]]
  then
    wait_pidfile $pidfile 1 $timeout $force
  else
    # TODO assume $1 is something to grep from 'ps ax'
    pid="$(ps auwwx | grep "$1" | awk '{print $2}')"
    wait_pid $pid 1 $timeout $force
  fi
}

check_nfs_mount() {
  opts=$1
  exports=$2
  mount_point=$3

  if grep -qs $mount_point /proc/mounts; then
    echo "Found NFS mount $mount_point"
  else
    echo "Mounting NFS..."
    mount $opts $exports $mount_point
    if [ $? != 0 ]; then
      echo "Cannot mount NFS from $exports to $mount_point, exiting..."
      exit 1
    fi
  fi
}

prep_system() {
  # bump maximum connections for higher TCP backlog setting
  echo 512 > /proc/sys/net/core/somaxconn || true

  # enable memory overcommit to quell warnings in redis startup log
  sysctl vm.overcommit_memory=1 || true

  # disable Transparent Huge Pages (THP) for redis' malloc discipline
  echo never > /sys/kernel/mm/transparent_hugepage/enabled || true
}
